### PR TITLE
Add assembly name to the private types when dumping dependencies

### PR DESCRIPTION
**Pull request by radekdoulik**
So that it can be used in the linker analyzer as a key to the type
sizes dictionary in the space analyzer.

The `IsAssemblyBound` method checks whether a type is visible outside
of an assembly. We need it for classes like
`<PrivateImplementationDetails>/__StaticArrayInitTypeSize=120`.

In this case the class hierarchy looks like this:

```
.class private auto ansi sealed '<PrivateImplementationDetails>'
       extends System.Object
{
 .class explicit ansi sealed nested private '__StaticArrayInitTypeSize=120'
        extends System.ValueType
 {
   ...
 }
}
```

### CODE REVIEW

File: linker/Mono.Linker/Annotations.cs [hunk 1]
```diff
@@ -349,11 +349,40 @@ public void Pop ()
...
+
+			return true;
+		}
+
 		string TokenString (object o)
 		{
 			if (o == null)
 				return "N:null";
 
+			if (o is TypeReference) {
```
Comment (marek-safar) [inline linker/Mono.Linker/Annotations.cs]: `o is TypeReference t` will give you `t` immediately

File: linker/Mono.Linker/Annotations.cs [hunk 2]
```diff
@@ -349,11 +349,40 @@ public void Pop ()
 			dependency_stack.Pop ();
 		}
 
+		bool IsNestedInPrivate (TypeDefinition td)
```
Comment (marek-safar) [inline linker/Mono.Linker/Annotations.cs]: I am not sure what is the intention here, what is nested in private?
You should also move td.IsNestedPrivate check here to make this work with 2+ levels of nesting

File: linker/Mono.Linker/Annotations.cs (hunk 1 above)
Comment (radekdoulik) [reply linker/Mono.Linker/Annotations.cs]: I knew there's faster way, forgot to check it before pushing. Thanks!

File: linker/Mono.Linker/Annotations.cs (hunk 2 above)
Comment (radekdoulik) [reply linker/Mono.Linker/Annotations.cs]: I have added explanation to the commit message and updated the patch.

**Pull request merged**